File: 2024/icfp/translator.py
#!/usr/bin/env python3

import unittest

__icfp_ascii = 'abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789!"#$%&\'()*+,-./:;<=>?@[\\]^_`|~ \n'
__icfp_index = {}

# Initialize __icfP_index
if len(__icfp_index) == 0:
    for i, c in enumerate(__icfp_ascii):
        __icfp_index[c] = i


def icfp_to_ascii(icfp):
    ascii = ''
    for c in icfp:
        ascii += __icfp_ascii[ord(c) - ord('!')]
    return ascii


def ascii_to_icfp(ascii):
    icfp = ''
    for c in ascii:
        icfp += chr(__icfp_index[c] + ord('!'))
    return icfp


def ascii_to_int(icfp):
    value = 0
    for c in icfp:
        value = value * 94 + __icfp_index[c]
    return value


# May be never used.
def int_to_ascii(value):
    if value == 0:
        return '!'

    ascii = ''
    while value > 0:
        ascii = __icfp_ascii[value % 94] + ascii
        value //= 94
    return ascii


def icfp_to_int(icfp):
    value = 0
    for c in icfp:
        value = value * 94 + ord(c) - ord('!')
    return value


def int_to_icfp(value):
    if value == 0:
        return '!'

    ascii = ''
    while value > 0:
        ascii = chr(value % 94 + ord('!')) + ascii
        value //= 94
    return ascii


def translate(message):
    return 'S' + ascii_to_icfp(message)


class TestTranslate(unittest.TestCase):
    def test_I_code(self):
        self.assertEqual(icfp_to_int('!'), 0)
        self.assertEqual(icfp_to_int('/6'), 1337)  # from the rules
        self.assertEqual(int_to_icfp(0), '!')
        self.assertEqual(int_to_icfp(1337), '/6')


if __name__ == '__main__':
    unittest.main()

File: 2024/icfp_rs/src/translator.rs
type Integer = i64;

static ICFP_ASCII: &str = &"abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789!\"#$%&'()*+,-./:;<=>?@[\\]^_`|~ \n";
static mut ICFP_INDEX: [usize; 256] = [0; 256];

// Initialize ICFP_INDEX
fn init_index() {
    unsafe {
        if ICFP_INDEX.len() == 0 {
            for (i, c) in ICFP_ASCII.chars().enumerate() {
                ICFP_INDEX[c as usize] = i;
            }
        }
    }
}

pub fn icfp_to_ascii(icfp: &str) -> String {
    let mut ascii = String::new();
    for c in icfp.chars() {
        ascii.push(ICFP_ASCII.chars().nth(c as usize - '!' as usize).unwrap());
    }
    ascii
}

pub fn ascii_to_icfp(ascii: &str) -> String {
    init_index();
    let mut icfp = String::new();
    unsafe {
        for c in ascii.chars() {
            icfp.push((ICFP_INDEX[c as usize] + '!' as usize) as u8 as char);
        }
    }
    icfp
}

pub fn ascii_to_int(icfp: &str) -> Integer {
    init_index();
    let mut value = 0;
    unsafe {
        for c in icfp.chars() {
            value = value * 94 + ICFP_INDEX[c as usize] as i64;
        }
    }
    value
}

// May be never used.
pub fn int_to_ascii(mut value: Integer) -> String {
    if value == 0 {
        return "!".to_string();
    }

    let mut ascii = String::new();
    while value > 0 {
        ascii.push(ICFP_ASCII.chars().nth((value % 94) as usize).unwrap());
        value /= 94;
    }
    ascii.chars().rev().collect()
}

pub fn icfp_to_int(icfp: &str) -> Integer {
    let mut value = 0;
    for c in icfp.chars() {
        value = value * 94 + c as Integer - '!' as Integer;
    }
    value
}

pub fn int_to_icfp(mut value: Integer) -> String {
    if value == 0 {
        return "!".to_string();
    }

    let mut ascii = String::new();
    while value > 0 {
        ascii.push((value % 94 + '!' as Integer) as u8 as char);
        value /= 94;
    }
    ascii.chars().rev().collect()
}

pub fn translate(message: &str) -> String {
    "S".to_string() + &ascii_to_icfp(message)
}

#[test]
fn test_integer_code() {
    assert_eq!(icfp_to_int("!"), 0);
    assert_eq!(icfp_to_int("/6"), 1337); // from the rules
    assert_eq!(int_to_icfp(0), "!");
    assert_eq!(int_to_icfp(1337), "/6");
}
